controllers/layered_image_handler.go

```diff
@@ -0,0 +1,164 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	ignTypes "github.com/coreos/ignition/v2/config/v3_2/types"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	LayeredImageDeployCm = "layered-image-deploy-cm"
+	image_mc_name        = "50-enable-sandboxed-containers-image"
+)
+
+// Process the LayeredImageDeployment feature gate (FG)
+// This method will be called by the processFeatureGates method during the beginning of the reconcile loop
+// The method will check for any existing MachineConfig related to KataConfig and return to the caller
+// by setting r.ImgMc to the image MachineConfig if present. This will allow the remainder of the reconcile loop
+// to use the image MachineConfig as needed.
+// If neither extension or image MachineConfig exists, and layeredImageDeployment feature is enabled,
+// then this method will create the image MachineConfig from the fg ConfigMap and set r.ImgMc to the
+// newly created image MachineConfig.
+// If layeredImageDeployment feature is disabled, then the method will reset r.ImgMc to nil
+// The key design aspect of this FG is that it has effect only during the creation of the KataConfig.
+// After creation of the KataConfig this FG has no effect.
+func (r *KataConfigOpenShiftReconciler) handleLayeredImageDeploymentFeature(state FeatureGateState) error {
+
+	// Check if MachineConfig exists and return the same without changing anything
+	mc, err := r.getExistingMachineConfig()
+	if err != nil {
+		r.Log.Info("Error in getting existing MachineConfig", "err", err)
+		return err
+	}
+
+	if mc != nil {
+		r.Log.Info("MachineConfig is already present. No changes will be done")
+		// If the MachineConfig is imageMachineConfig, then set r.ImgMc to the same
+		if mc.Name == image_mc_name {
+			r.ImgMc = mc
+		}
+		return nil
+	}
+
+	if state == Enabled {
+		r.Log.Info("LayeredImageDeployment feature is enabled")
+
+		cm := &corev1.ConfigMap{}
+		err := r.Client.Get(context.Background(), types.NamespacedName{
+			Name:      LayeredImageDeployCm,
+			Namespace: OperatorNamespace,
+		}, cm)
+		if err != nil {
+			r.Log.Info("Error in retrieving LayeredImageDeployment ConfigMap", "err", err)
+			return err
```

Comment:
Should we here again handle "not found" separately from other errors?

If the error is other than "not found" then it likely means just an (probably) intermittent failure to talk to the control plane.

If it's "not found" (and if I understand the logic of this PR correctly) then it's a user error which we should try to communicate to the user ("you enabled image-based deployment but gave no image URL and we can't work with that") _unless_ we also want to provide a default image URL.  In that case, I think we should proceed as if we found a configmap with the default image URL and empty kernel args.